arctic_training/trainer/dpo_trainer.py

```diff
@@ -0,0 +1,174 @@
+from arctic_training.registry import register
+from arctic_training.scheduler.hf_factory import HFSchedulerFactory
+from arctic_training.tokenizer.hf_factory import HFTokenizerFactory
+from arctic_training.trainer.trainer import Trainer
+from arctic_training.config.trainer import TrainerConfig
+
+
+
+def to_device(batch: Dict, device: str) -> Dict:
+    output = {}
+    for k, v in batch.items():
+        if isinstance(v, torch.Tensor):
+            output[k] = v.to(device)
+    return output
+
+
+
+def get_logprobs(
+        logits: torch.Tensor, labels: torch.Tensor,
+        ignore_label_index: int
+    ) -> Tuple[float, float]:
+    """Compute the log probabilities of the given labels under the given logits.
+
+    Args:
+        logits: Logits of the model (unnormalized). Shape: (batch_size, sequence_length, vocab_size)
+        labels: Labels for which to compute the log probabilities. Label tokens with a value of label_pad_token_id are ignored. Shape: (batch_size, sequence_length)
+        label_pad_token_id: The label pad token id.
+        is_encoder_decoder: Whether the model is an encoder-decoder model.
+
+    Returns:
+        A Tuple of two tensor of shape ((batch_size,), (batch_size,)) containing the sum of log probabilities of the given labels under the given logits in the first tensor and the number of non-masked tokens in the second tensor.
+    """
+    if logits.shape[:-1] != labels.shape:
+        raise ValueError(
+            f"Logits (batch and sequence length dim) {logits.shape[:-1]} and labels must have the same shape {labels.shape}."
+        )
+
+    labels = labels[:, 1:].clone()
+    logits = logits[:, :-1, :]
+    loss_mask = labels != ignore_label_index
+
+    # dummy token; we'll ignore the losses on these tokens later
+    labels[labels == ignore_label_index] = 0
+
+    per_token_logps = torch.gather(
+        logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)
+    ).squeeze(2)
+
+    return (per_token_logps * loss_mask).sum(-1), loss_mask.sum(-1)
+
+class DPOTrainerConfig(TrainerConfig):
+    type: str = "dpo"
+    ref_model: ModelConfig
+    beta: float
+    ignore_label_index: int = -100
+    label_smoothing: float = 0.0
+    """ Model configuration. """
+
+
+class DPOTrainer(Trainer):
+    name = "dpo"
+    config: DPOTrainerConfig
+    model_factory: Union[HFModelFactory, LigerModelFactory]
+    ref_model_factory: Union[HFModelFactory, LigerModelFactory]
+    checkpoint_engine: Union[DSCheckpointEngine, HFCheckpointEngine]
+    optimizer_factory: FusedAdamOptimizerFactory
+    scheduler_factory: HFSchedulerFactory
+    tokenizer_factory: HFTokenizerFactory
+    callbacks = [("post-init", init_ref_model)]
+
+
+    def init_ref_model(self):
+        ref_model_factory = self.config.ref_model.factory(trainer=self, config=self.config.ref_model) # Be explicit about which model config to use
+        self.ref_model = ref_model_factory()
```

Comment:
```suggestion
def init_ref_model(self: Trainer):
    ref_model_factory = self.config.ref_model.factory(trainer=self, config=self.config.ref_model) # Be explicit about which model config to use
    self.ref_model = ref_model_factory()

class DPOTrainer(Trainer):
    name = "dpo"
    config: DPOTrainerConfig
    model_factory: Union[HFModelFactory, LigerModelFactory]
    ref_model_factory: Union[HFModelFactory, LigerModelFactory]
    checkpoint_engine: Union[DSCheckpointEngine, HFCheckpointEngine]
    optimizer_factory: FusedAdamOptimizerFactory
    scheduler_factory: HFSchedulerFactory
    tokenizer_factory: HFTokenizerFactory
    callbacks = [("post-init", init_ref_model)]
```

`init_ref_model` needs to be defined outside `DPOTrainer` or we will not know what this callback is referencing. The other option is to make it a static method of `DPOTrainer` and set the callback to `("post-init", DPOTrainer.init_ref_model)`